### DISCUSSION TITLE
Alignment changes in some of the pdf file after replace text and replaced text is hidden

### DISCUSSION THREAD

**Discussion by jayakumar0612** [pymupdf/PyMuPDF#1997] [Looking for help] [answered]
Hi, 
I'm trying to find and replace text using below code

    doc = fitz.open(input_file_path)
            dict.pop("source_dict")
            dict.pop('irrelevant_fields')
            origin = ""
             fsize = ""
            font = ""

        for field in dict:
            for value in dict[field]:
                for page in doc:
                    search_text_list = [" "+value+" ", " "+value+",", " "+value+"."]
                    for search_text in search_text_list:
                        rl = page.search_for(search_text)
                        replacement_text = "{{" + str(field) + "}}"
                        #replacement_text = "*" * len(field)
                        for rect in rl:
                            for b in page.get_text("dict", clip=rect)["blocks"]:
                                for l in b["lines"]:
                                    for span in l["spans"]:
                                        fsize = span["size"]
                                        origin = fitz.Point(span["origin"])
                                        flags = span["flags"]
                                        try:
                                            font = page.get_text("dict")["blocks"][0]["lines"][0]['spans'][0]['font']
                                        except Exception:
                                            if flags & 2 ** 3:
                                                font = "cour"
                                            else:
                                                font = "helv"
                                        if not font:
                                            if flags & 2 ** 3:
                                                font = "cour"
                                            else:
                                                font = "helv"
                            page.add_redact_annot(rect)
                            page.apply_redactions()

                            # tl = fitz.get_text_length(search_text, fontname=font, fontsize=fsize)
                            # fsize = fsize * rect.width / tl
                            if len(search_text) < len(replacement_text):
                                fsize = fsize / 4
                            if origin and font and fsize:
                                page.insert_text(origin, replacement_text, fontname=font, fontsize=fsize, overlay=False)

        path = os.path.split(input_file_path)[1]
        if os.path.exists(path):
            os.remove(path)
        doc.save(path,garbage=3, deflate=True)
        doc.close()

Some files it is working fine but for some files the replaced text is not visible when opening the file after replace. and when we search for a term it highlights some ir-releant term in pdf viewer software.  i have attached source and replaced file for

**Comment by JorjMcKie** [comment]
I did not try out your files and the code yet, but this sounds like one of the typical issues where pages contain "unprotected" addressing scheme changes - see the docs [here](https://pymupdf.readthedocs.io/en/latest/recipes-common-issues-and-their-solutions.html#misplaced-item-insertions-on-pdf-pages).
So a solution may be to immediately after a page has been loaded, do `page.clean_contents()`. This will clean the page's appearance building instructions and thus take care of correctly positioning new insertions of text and other stuff.

**Reply by jayakumar0612** [reply]
Hi @JorjMcKie  Thanks for the quick response. I tried to do page.clean_contents() but it is not solving the issue. As per [[doc  ](https://pymupdf.readthedocs.io/en/latest/recipes-common-issues-and-their-solutions.html#misplaced-item-insertions-on-pdf-pages)](url)
the issue is not placed at a different location. it has replaced at the expected location but the replaced text is not visible. the replaced area looks blank. but, if i copy the blank area and paste it i a notepad i can get the  text that is inserted.
So the issue is replaced text is hidden or not visible for some files.

**Reply by JorjMcKie** [reply]
AH I just saw you are doing overlay=False!
This puts the text **behind** the page's other content! Use overlay=True or let it default.

**Reply by jayakumar0612** [reply]
I have changed overlay=True but the inserted text shows something like this

![image](https://user-images.githubusercontent.com/116795686/198364711-c9edcdde-94e7-4a8d-95c4-fae1f938b1c5.png)

**Reply by JorjMcKie** [reply]
> I have changed overlay=True but the inserted text shows something like this
> 
> ![image](https://user-images.githubusercontent.com/116795686/198364711-c9edcdde-94e7-4a8d-95c4-fae1f938b1c5.png)

instead of what?

**Reply by jayakumar0612** [reply]
instead of the inserted word "ORGANIZATION" the above code is shown

**Reply by JorjMcKie** [reply]
Hm, I don't have your full script. So please tell me which word on which page you want to replace with "ORGANIZATION" and will try to reproduce the behavior.

**Reply by jayakumar0612** [reply]
I'm trying to replace word  "Gratuity Act" in below image with "ORGANIZATION"
![image](https://user-images.githubusercontent.com/116795686/198819274-932bd195-2b66-49db-9628-ed6ed1af13eb.png)
the below image shows after replacement with overlay=False here "ORGANIZATION" is hidden
![image](https://user-images.githubusercontent.com/116795686/198819312-aff64561-597b-47a5-863a-efc58a0358aa.png)
below image shows after replacement with overlay=True here "ORGANIZATION" is shown as code.
![image](https://user-images.githubusercontent.com/116795686/198819401-69d1f0ff-5852-4d46-827e-71c343461910.png)

I have attached two pdf in my first post. one is original file and another one is after replacement with overlay=False

I think The below code snippet should be helpful to reproduce the issue

    def replace_text():
      import fitz
      import os
  
      input_file_path = "replace path to pdf file here"
      doc = fitz.open(input_file_path)
      origin = ""
      fsize = ""
      font = ""
      value = "Gratuity Act"
      field = "ORGANIZATION"
      
  
      for page in doc:
          search_text_list = [" " + value + " ", " " + value + ",", " " + value + "."]
          for search_text in search_text_list:
              rl = page.search_for(search_text)
              replacement_text = "{{" + str(field) + "}}"
              # replacement_text = "*" * len(field)
              for rect in rl:
                  for b in page.get_text("dict", clip=rect)["blocks"]:
                      for l in b["lines"]:
                          for span in l["spans"]:
                              fsize = span["size"]
                              origin = fitz.Point(span["origin"])
                              flags = span["flags"]
                              try:
                                  font = page.get_text("dict")["blocks"][0]["lines"][0]['spans'][0]['font']
                              except Exception:
                                  if flags & 2 ** 3:
                                      font = "cour"
                                  else:
                                      font = "helv"
                              if not font:
                                  if flags & 2 ** 3:
                                      font = "cour"
                                  else:
                                      font = "helv"
                  page.add_redact_annot(rect)
                  page.apply_redactions()
  
                  # tl = fitz.get_text_length(search_text, fontname=font, fontsize=fsize)
                  # fsize = fsize * rect.width / tl
                  if len(search_text) < len(replacement_text):
                      fsize = fsize / 4
                  if origin and font and fsize:
                      page.insert_text(origin, replacement_text, fontname=font, fontsize=fsize, overlay=False)
  
      path = os.path.split(input_file_path)[1]
      if os.path.exists(path):
          os.remove(path)
      doc.save(path, garbage=3, deflate=True)
      doc.close()

**Comment by JorjMcKie** [comment, marked as answer]
Something must be wrong with your script. Here is what I did:
```python
page=doc[0]
rl=page.search_for(stext)

spans=[]
for r in rl:
    for b in page.get_text("dict",clip=r)["blocks"]:
        for l in b["lines"]:
            for s in l["spans"]:
                spans.append(s)

page.apply_redactions()

for s in spans:
    bbox=s["bbox"]
    fsize=s["size"]
    origin=s["origin"]
    page.insert_text(origin, text,fontsize=fsize)

doc.save("x.pdf")
```
And it does work!
The following also work, but the text is fairly small because it had to be shrinked to fit the old place.
```python
page=doc[0]
page.clean_contents()
for r in rl:
    page.add_redact_annot(r, text)

page.apply_redactions()

doc.save("y.pdf")
```
In neither case appears your funny text.

**Reply by jayakumar0612** [reply]
Thanks for your support @JorjMcKie  the above code snippet works fine for me
